Adds in form validation mixin, error reporting on inputs, and a few examples

Sean-McQ opened the pull request:
Adds in form validation mixin for use with cru-resource mixin to validate values in form payload and provide feedback to the user.

Also adds in rule based validation to labeled input fields, fields will display error messages in a rule returns a message, validated in real time.

Specific instructions for both included in developer docs, a few example forms have implemented the mixin and field "rules" property as an example, and a couple additional components have been implemented and/or extended to allow for more robust error messaging to users.

catherineluse commented:
In the design review meeting today, it was said that this needs to be easy to use.

I think we need to be more specific about what we mean by "easy to use." As in, what exactly should be the API for adding the following:

- A new custom validator
- Adding form validation to a new field of a form that already has some form validation
- Adding form validation to a form that doesn't have any validation already

@Sean-McQ in terms of step-by-step instructions, how many steps would it take to do each of those 3 things? If it has just a couple steps I would say that counts as easy to use.

cnotv commented on shell/components/form/Error.vue:
```diff
@@ -0,0 +1,37 @@
+<script>
+export default {
+  props: {
+    value: {
```
Please add jsdoc annotation for new components.

Sean-McQ replied:
This is a pretty straightforward component and I'm not certain we've established a standard for commenting every prop on new components just yet. I'll keep this in mind for more complex components in the future though.

cnotv commented on shell/components/form/Error.vue:
```diff
@@ -0,0 +1,37 @@
+    return {};
+  },
+  computed: {
+    validationMessage() {
```
Could you please make a test out of this?

As far as I can see we are pushing the result of a function, which can be pretty much anything.
I would recommend to add specificity with a type.
You must consider also that if you may return `undefined` or empty string, you may end up with results as the following:
```ts
[undefined, ''].join(', ')      //      ', '
```

cnotv commented on shell/components/Tabbed/index.vue:
```diff
@@ -230,7 +230,7 @@ export default {
         >
           <span>{{ tab.labelDisplay }}</span>
-          <i v-if="tab.displayAlertIcon" class="conditions-alert-icon icon-error icon-lg" />
+          <i v-if="tab.displayAlertIcon || (tab.error && !tab.active)" v-tooltip="t('validation.tab')" class="conditions-alert-icon icon-warning icon-lg" />
```
Could we abstract the logic and avoid it in the markup?
E.g. 
```ts
methods: {
// ...
  hasIcon(tab: unknown): boolean {
    return tab.displayAlertIcon || (tab.error && !tab.active)
  }
// ...
}
```

Sean-McQ replied:
Yeah, this one's getting a little long in the tooth.

cnotv commented on shell/components/form/InputWithSelect.vue:
```diff
@@ -75,6 +75,14 @@ export default {
       type:    String,
       default: '',
     },
+    textRules: {
```
We should have just one set of rules, regardless of the adopted type.

Sean-McQ replied:
I don't know if I agree with that. For some of the more complex sub-forms it makes sense to pass in rulesets as a subselection of the parent form's rule sets and then parse it to route the correct rulesets to the correct fields and there's methods in the mixin to do that when required but for something simple like this one that seems like overkill, especially given the amount of extra logic required to do so (given that this sub-form in particular is essentially just a mashup of two fields with some styling). In some cases, that's probably the right move but here I don't think the juice is worth the squeeze.

cnotv commented on shell/components/form/NameNsDescription.vue:
```diff
@@ -140,6 +140,18 @@ export default {
       type:    Boolean,
       default: true,
+    },
+    namespaceRules: {
```
There's cases where we have even more inputs, maybe we pass an object with the key of the model?

```ts
rules: {
  default: () => {
    namespace: [],
    name: [],
    description: []
  },
  type: Object,  // We can convert to TS another time I guess
}
```

We should probably create a global dictionary type for these cases.
```ts
type Ruleset = (value: unknown) => string | undefined
type RulesetsComponent = Record<string, Ruleset[]>
```

cnotv commented on shell/components/form/ServicePorts.vue:
```diff
@@ -27,6 +28,12 @@ export default {
+      default:   () => [],
+      type:      Array,
+      // we only want functions in the rules array
+      validator: rules => rules.every(rule => ['function'].includes(typeof rule))
```
I dont think is reliable and we should definitely do not duplicate code.

Sean-McQ replied:
I don't see the failure point but I agree it's cleaner in TS than with this function. Unfortunately, converting a component of this size to TS just to get cleaner prop validation is probably beyond the scope of this PR at this time.

cnotv reviewed (changes requested):
This is a first review, I'll have to continue next time.

rak-phillip commented on docs/developer/forms-and-validation.md:
```diff
@@ -49,17 +49,47 @@ If a form element was repeated for every row in a table, it would make the UI sl
+
+## Form Validation
+
+While individual fields are typically responsible for displaying their own validation messages, the form itself should also contain logic to disable the save button should validation errors exist, display errors not bound to a specific field, and/or display errors returned by the API. This is mostly accomplished via a mixin called `form-validation` (found at `shell/mixins/form-validation.js`). Ev ...
```
```suggestion
While individual fields are typically responsible for displaying their own validation messages, the form itself should also contain logic to disable the save button should validation errors exist, display errors not bound to a specific field, and/or display errors returned by the API. This is mostly accomplished via a mixin called `form-validation` (found at `shell/mixins/form-validation.js`). Eventually, this mixin will be embedded into the existing `CreateEditView` mixin already used by every form but is currently separate to facilitate a safer implementation of the new Form Validation logic.
```

rak-phillip replied:
1. Missing comma after Eventually
2. separate is misspelled

rak-phillip commented on docs/developer/forms-and-validation.md:
```diff
@@ -49,17 +49,47 @@ If a form element was repeated for every row in a table, it would make the UI sl
+2. If the validation is passed, the function returns `undefined`.
+3. If the validation fails, the function returns a string to be rendered by the field component as an error.
+
+In addition to the "rules" parameter existing on the LabeledInput and LabeledSelect form components, an additional component called Error (found at `shell/components/form/Error`) exists which can also accept a value and rules. This allows the developer to display an error message which may not be tied to a specific field or tied to a field that is somehow incapable of displaying it's own validati ...
```
```suggestion
In addition to the "rules" parameter existing on the LabeledInput and LabeledSelect form components, an additional component called Error (found at `shell/components/form/Error`) exists which can also accept a value and rules. This allows the developer to display an error message which may not be tied to a specific field or tied to a field that is somehow incapable of displaying its own validation message.
```

rak-phillip commented on docs/developer/forms-and-validation.md:
```diff
@@ -49,17 +49,47 @@ If a form element was repeated for every row in a table, it would make the UI sl
+2. Additional rules to be validated by the form are also pulled directly from the resource model's `modelValidationRules` property which converts the model's customValidationRules into a format useable by the `form-validation` mixin. The combined `fvFormRuleSets` and the `modelValidationRules` can be viewed in a single array under the computed property `rulesets` provided by the mixin.
+3. In order to avoid duplication of validator definitions, it's recommended to pull the function array for a field's "rules" property directly from the components `fvFormRuleSets`. The `form-validation` mixin provides two helper methods for this named `fvGetPathRules` and `fvGetAndReportPathRules`, the specifics as to which one to use and when are defined in the comments in the mixin file itself.
+
+*It is highly recommended to read through the comments in the `form-validation` mixin to understand it's various properties and their suggested use-cases.
```
```suggestion
*It is highly recommended to read through the comments in the `form-validation` mixin to understand its various properties and their suggested use-cases.
```

rak-phillip commented on shell/components/Tabbed/Tab.vue:
```diff
@@ -32,6 +32,11 @@ export default {
       type:    Boolean,
       default: null
     },
+    // ToDo: this is redundant with the above, fix that...
```
Can we make sure to have an associated issue mentioned inline with the todo?

catherineluse replied:
I agree. I think we shouldn't have any ToDos in the code itself because the GitHub issues are more organized.

Sean-McQ replied:
Looking at this closer, these two props aren't perfectly redundant. `displayAlertIcon` will always show an alert icon whereas `error` will only show an error if the tab is not currently active. Removing the ToDo for now.

rak-phillip reviewed (changes requested):
I wasn't able to make it all the way through, but I did encounter an issue while creating a new RKE2 cluster in Digital Ocean. It looks like an error is thrown after clicking the Create button without entering any input:

```
TypeError: can't access property "$store", this is undefined
    t i18n.js:28
    required index.js:64
    validationMessage labeled-form-element.ts:131
    VueJS 4
    render LabeledInput.vue:127
    VueJS 14
    onBlurLabeled labeled-form-element.ts:184
    onBlur LabeledInput.vue:224
    VueJS 33
```

This issue isn't present in v2.6-head

catherineluse commented on docs/developer/forms-and-validation.md:
```diff
@@ -49,17 +49,47 @@ If a form element was repeated for every row in a table, it would make the UI sl
+
+Existing validation functions can be found at `shell/utils/validators/formRules` in the dashboard codebase but it is possible to create custom functions to accomplish the same purpose if need be. A form rule validator must meet the following criteria in order to be useful for this purpose:
+
+1. The function must accept a single parameter which will be the value of the field to be validated.
```
I think it would help if this doc linked to an example in the code

catherineluse commented on docs/developer/forms-and-validation.md:
```diff
@@ -49,17 +49,47 @@ If a form element was repeated for every row in a table, it would make the UI sl
+
+The `form-validation` mixin itself includes most of the information a developer will need to use it in comments in the file itself but a high-level summary would cover the following points:
+
+1. The specific rules to be validated by the form should be located in a data property for the component named `fvFormRuleSets` which is an array of objects defining the specific rulesets for specific paths in the forms value to validate. The specific properties of the objects required by this array are defined in the mixin file itself as comments.
```
I think it would help to have a minimalistic code example below these steps. Basically the simplest possible Vue component that uses this mixin

catherineluse reviewed (changes requested):
Something is wrong with the Storage tab in the create/edit workload form. When I selected a storage type it crashed:
<img width="1198" alt="Screen Shot 2022-06-27 at 11 55 06 AM" src="https://user-images.githubusercontent.com/20599230/176015152-be114f7d-3f4e-4704-b11f-a1e4329e2007.png">
<img width="875" alt="Screen Shot 2022-06-27 at 11 55 12 AM" src="https://user-images.githubusercontent.com/20599230/176015197-b05fd7a5-fb1b-4c0f-b47f-c3bdf3a8ff45.png">

catherineluse commented:
In the Pod Scheduling tab, the Save button wasn't disabled even though the required Topology Key field was empty:
<img width="1393" alt="Screen Shot 2022-06-27 at 12 02 09 PM" src="https://user-images.githubusercontent.com/20599230/176016286-703217fb-18f8-40e8-8212-fd2409c12f3f.png">

Sean-McQ commented on shell/components/form/Error.vue:
```diff
@@ -0,0 +1,37 @@
+      default:   () => [],
+      type:      Array,
+      // we only want functions in the rules array
+      validator: rules => rules.every(rule => ['function'].includes(typeof rule))
```
> can't access property "$store", this is undefined

Took me a while to dig into, best guess at the moment is that this happened when I converted the component mixin over to TS because it wasn't doing that before. It's a little annoying because I'd like to provide the required message as well as the asterix on the field even when it's marked as required using the old method but it's not really required for this PR and it's probably best to eventually sunset the "required" prop on labeled-form-elements in favor of the newer form validation anyways. So for now I've removed the offending bit of code which doesn't break anything that was already there but also doesn't add in the backwards compatibility for explicit error messages on the field for required inputs.

Sean-McQ commented:
> I wasn't able to make it all the way through, but I did encounter an issue while creating a new RKE2 cluster in Digital Ocean. It looks like an error is thrown after clicking the Create button without entering any input:
> 
> ```
> TypeError: can't access property "$store", this is undefined
>     t i18n.js:28
>     required index.js:64
>     validationMessage labeled-form-element.ts:131
>     VueJS 4
>     render LabeledInput.vue:127
>     VueJS 14
>     onBlurLabeled labeled-form-element.ts:184
>     onBlur LabeledInput.vue:224
>     VueJS 33
> ```
> 
> This issue isn't present in v2.6-head

Took me a while to dig into, best guess at the moment is that this happened when I converted the component mixin over to TS because it wasn't doing that before. It's a little annoying because I'd like to provide the required message as well as the asterix on the field even when it's marked as required using the old method but it's not really required for this PR and it's probably best to eventually sunset the "required" prop on labeled-form-elements in favor of the newer form validation anyways. So for now I've removed the offending bit of code which doesn't break anything that was already there but also doesn't add in the backwards compatibility for explicit error messages on the field for required inputs.

Sean-McQ commented:
> Something is wrong with the Storage tab in the create/edit workload form. When I selected a storage type it crashed: <img alt="Screen Shot 2022-06-27 at 11 55 06 AM" width="1198" src="https://user-images.githubusercontent.com/20599230/176015152-be114f7d-3f4e-4704-b11f-a1e4329e2007.png"> <img alt="Screen Shot 2022-06-27 at 11 55 12 AM" width="875" src="https://user-images.githubusercontent.com/20599230/176015197-b05fd7a5-fb1b-4c0f-b47f-c3bdf3a8ff45.png">
> 
> Similar to Phil, I also had errors when trying to create an RKE2 cluster in Digital Ocean: <img alt="Screen Shot 2022-06-27 at 11 58 29 AM" width="1378" src="https://user-images.githubusercontent.com/20599230/176015646-28ce011e-141d-4280-abad-0f331d957504.png">

I can't reproduce this after the fix to the issue that Phil brought up

Sean-McQ commented on pkg/rancher-components/src/components/Form/LabeledInput/LabeledInput.vue:
```diff
@@ -301,6 +301,11 @@ export default (
       :value="tooltip"
       :status="status"
     />
+    <LabeledTooltip
```
I don't think we can avoid a bunch of logic here but we can move some around to make it a bit less verbose. We still end up with two LabeledTooltips in the markup which I'm okay with since they serve two separate purposes and theoretically we might want to render both at the same time at some point. I did however combine the two LabeledTooltips that were already there since they were basically doing the same thing.

Sean-McQ commented:
The crashes that @rak-phillip and @catherineluse should all be fixed.
